File: UltraMesh 2d animation/Core/Tools/BoneTool.swift
import Foundation
import simd

final class BoneTool: Tool {
    let type: ActiveTool = .bone

    private enum Interaction {
        case create(parentID: UUID?, start: SIMD2<Float>)
        case moveRoot(UUID)
        case moveTip(UUID)
    }

    private enum HitPart {
        case start
        case end
        case segment
    }

    private var interaction: Interaction?
    private var currentPreview: SIMD2<Float>?
    private var mouseDownPosition: SIMD2<Float>?
    private var moveThresholdExceeded = false

    #if os(iOS)
    // iOS: larger targets for comfortable finger + Pencil use.
    // Drawable pixels; at 2× display scale: joint≈26pt, segment≈16pt.
    private let jointRadius: Float = 52
    private let lineRadius: Float  = 32
    private let clickDragThreshold: Float = 8
    #else
    private let jointRadius: Float = 12
    private let lineRadius: Float  = 9
    private let clickDragThreshold: Float = 4
    #endif

    func onMouseDown(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        scene.beginInteraction()
        _ = assets
        if let hit = hitTestBonePart(
            screenPoint: input.screenPosition,
            viewSize: input.viewSize,
            scene: scene,
            camera: input.camera
        ) {
            if input.isCommandPressed || input.isShiftPressed {
                // Cmd-click toggles the bone in the multi-selection without
                // starting a drag — matches macOS finder/list conventions.
                scene.toggleBoneSelection(hit.id)
                interaction = nil
                currentPreview = nil
                scene.setBoneCreationPreview(start: nil, end: nil)
                return
            }
            scene.selectBone(hit.id)
            currentPreview = nil
            interaction = interactionForHit(hit, input: input)
            mouseDownPosition = input.position
            moveThresholdExceeded = false
            if case let .create(_, start) = interaction {
                currentPreview = input.position
                scene.setBoneCreationPreview(start: start, end: input.position)
            } else {
                scene.setBoneCreationPreview(start: nil, end: nil)
            }
            return
        }

        let chainParentID = scene.selectedBoneID
        interaction = .create(parentID: chainParentID, start: input.position)
        currentPreview = input.position
        mouseDownPosition = input.position
        moveThresholdExceeded = true  // create mode always tracks drag from frame 0
        scene.setBoneCreationPreview(start: input.position, end: input.position)
    }

    func onMouseDrag(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        _ = assets
        guard let interaction else { return }
        switch interaction {
        case let .create(_, start):
            currentPreview = input.position
            scene.setBoneCreationPreview(start: start, end: input.position)
        case let .moveRoot(boneID):
            guard ensureDragStarted(currentPosition: input.position) else { return }
            scene.moveBoneRoot(id: boneID, to: input.position)
        case let .moveTip(boneID):
            guard ensureDragStarted(currentPosition: input.position) else { return }
            scene.moveBoneTip(id: boneID, to: input.position)
        }
    }

    // Returns true once the cursor has moved beyond the click/drag threshold
    // since mouseDown. Prevents micro-jitter on a click from being interpreted
    // as a drag that teleports a bone joint.
    private func ensureDragStarted(currentPosition: SIMD2<Float>) -> Bool {
        if moveThresholdExceeded { return true }
        guard let down = mouseDownPosition else { return false }
        if simd_distance(currentPosition, down) >= clickDragThreshold {
            moveThresholdExceeded = true
            return true
        }
        return false
    }

    func onMouseUp(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        scene.endInteraction()
        _ = assets
        defer {
            interaction = nil
            currentPreview = nil
            mouseDownPosition = nil
            moveThresholdExceeded = false
            scene.setBoneCreationPreview(start: nil, end: nil)
        }

        guard let interaction else { return }
        guard case let .create(parentID, start) = interaction else { return }
        let end = currentPreview ?? input.position
        guard simd_distance(start, end) >= 8 else {
            // A CLICK ON EMPTY CANVAS, which is a press and a release that
            // never travelled far enough to be a bone. It used to be nothing at
            // all: no bone made, and nothing cleared either, so the next drag
            // still chained off whatever was selected before it and there was
            // no way to start a second root without going to the hierarchy.
            //
            // It breaks the chain now. The SAME condition decides both, so
            // "too short to be a bone" and "a click that cancels" cannot drift
            // apart into two thresholds that disagree.
            scene.selectBone(nil)
            return
        }
        scene.addBone(start: start, end: end, parentID: parentID)
    }

    private func interactionForHit(_ hit: (id: UUID, part: HitPart), input: ToolInput) -> Interaction? {
        switch hit.part {
        case .start, .segment:
            return .moveRoot(hit.id)
        case .end:
            // A TIP MEANS "CARRY ON FROM HERE".
            //
            // It used to mean "resize this bone", with Shift for chaining —
            // which put the one gesture you want after making a bone behind a
            // modifier, and made the obvious gesture do something else. Building
            // a run of short bones meant moving away from the tip, clicking
            // empty space and coming back for every one of them. The two are
            // swapped: Shift resizes, a plain drag chains.
            if input.isShiftPressed {
                return .moveTip(hit.id)
            }
            return .create(parentID: hit.id, start: input.position)
        }
    }

    private func hitTestBonePart(
        screenPoint: SIMD2<Float>,
        viewSize: SIMD2<Float>,
        scene: SceneManager,
        camera: CameraState?
    ) -> (id: UUID, part: HitPart)? {
        let segments = scene.skeleton.worldLineSegments()
        guard !segments.isEmpty else { return nil }

        // Joints and segments are tracked separately so joints always win when
        // both overlap — a segment at 8px never beats a joint at 12px.
        var bestJointHit: (id: UUID, part: HitPart)?
        var bestJointDist = Float.greatestFiniteMagnitude
        var bestSegHit:   (id: UUID, part: HitPart)?
        var bestSegDist   = Float.greatestFiniteMagnitude

        for (bone, start, end) in segments {
            let startScreen = camera?.worldToScreen(start, viewSize: viewSize) ?? (start + viewSize * 0.5)
            let endScreen   = camera?.worldToScreen(end,   viewSize: viewSize) ?? (end   + viewSize * 0.5)

            let startDist = simd_distance(screenPoint, startScreen)
            if startDist <= jointRadius && startDist < bestJointDist {
                bestJointDist = startDist
                bestJointHit  = (bone.id, .start)
            }

            let endDist = simd_distance(screenPoint, endScreen)
            if endDist <= jointRadius && endDist < bestJointDist {
                bestJointDist = endDist
                bestJointHit  = (bone.id, .end)
            }

            let lineDist = distancePointToSegment(point: screenPoint, a: startScreen, b: endScreen)
            if lineDist <= lineRadius && lineDist < bestSegDist {
                bestSegDist = lineDist
                bestSegHit  = (bone.id, .segment)
            }
        }

        // Joints take absolute priority — a joint hit always beats a segment hit.
        return bestJointHit ?? bestSegHit
    }

    private func distancePointToSegment(point: SIMD2<Float>, a: SIMD2<Float>, b: SIMD2<Float>) -> Float {
        let ab = b - a
        let t = max(0, min(1, simd_dot(point - a, ab) / max(0.0001, simd_dot(ab, ab))))
        let projection = a + ab * t
        return simd_distance(point, projection)
    }
}

File: UMeshCore/src/Editor/Tools/BoneTool.cpp
#include "umeshcore/Editor/Tools/BoneTool.h"

#include <limits>

namespace umeshcore {

BoneTool::Radii BoneTool::radiiFor(bool touchOptimized) {
    // iOS: larger targets for comfortable finger + Pencil use. Drawable
    // pixels; at 2x display scale: joint ~26pt, segment ~16pt.
    if (touchOptimized) return Radii{52.0f, 32.0f, 8.0f};
    return Radii{12.0f, 9.0f, 4.0f};
}

bool BoneTool::ensureDragStarted(const Vec2& currentPosition, float clickDragThreshold) {
    if (moveThresholdExceeded_) return true;
    if (!mouseDownPosition_.has_value()) return false;
    if (length(currentPosition - *mouseDownPosition_) >= clickDragThreshold) {
        moveThresholdExceeded_ = true;
        return true;
    }
    return false;
}

std::optional<BoneTool::Interaction> BoneTool::interactionForHit(const BoneHit& hit, const ToolInput& input) {
    if (hit.part == BoneHitPart::Start || hit.part == BoneHitPart::Segment) {
        Interaction i;
        i.kind = InteractionKind::MoveRoot;
        i.boneID = hit.id;
        return i;
    }
    // A tip means "carry on from here": a plain drag chains a new child
    // bone; Shift resizes the existing one instead. NOTE: as of the
    // current onMouseDown (which returns early on
    // isShiftPressed/isCommandPressed before ever reaching this function --
    // see BoneTool.h's file header), `isShiftPressed` is always false by
    // the time this runs, so this branch is presently unreachable. Ported
    // as-is rather than removed, since the intent is clear and removing it
    // would silently diverge further from the Swift source if that call
    // site's ordering is ever fixed upstream.
    if (input.isShiftPressed) {
        Interaction i;
        i.kind = InteractionKind::MoveTip;
        i.boneID = hit.id;
        return i;
    }
    Interaction i;
    i.kind = InteractionKind::Create;
    i.parentID = hit.id;
    i.start = input.position;
    return i;
}

std::optional<BoneHit> BoneTool::hitTestBonePart(
    const Vec2& screenPoint, const Vec2& viewSize, const Skeleton& skeleton, CameraState* camera,
    float jointRadius, float lineRadius) {
    const auto segments = skeleton.worldLineSegments();
    if (segments.empty()) return std::nullopt;

    // Joints and segments are tracked separately so joints always win when
    // both overlap -- a segment at 8px never beats a joint at 12px.
    std::optional<BoneHit> bestJointHit;
    float bestJointDist = std::numeric_limits<float>::max();
    std::optional<BoneHit> bestSegHit;
    float bestSegDist = std::numeric_limits<float>::max();

    for (const auto& seg : segments) {
        const Vec2 startScreen =
            camera != nullptr ? camera->worldToScreen(seg.start, viewSize) : seg.start + viewSize * 0.5f;
        const Vec2 endScreen =
            camera != nullptr ? camera->worldToScreen(seg.end, viewSize) : seg.end + viewSize * 0.5f;

        const float startDist = length(screenPoint - startScreen);
        if (startDist <= jointRadius && startDist < bestJointDist) {
            bestJointDist = startDist;
            bestJointHit = BoneHit{seg.bone.id, BoneHitPart::Start};
        }
        const float endDist = length(screenPoint - endScreen);
        if (endDist <= jointRadius && endDist < bestJointDist) {
            bestJointDist = endDist;
            bestJointHit = BoneHit{seg.bone.id, BoneHitPart::End};
        }
        const float lineDist = ToolUtilities::distancePointToSegment(screenPoint, startScreen, endScreen);
        if (lineDist <= lineRadius && lineDist < bestSegDist) {
            bestSegDist = lineDist;
            bestSegHit = BoneHit{seg.bone.id, BoneHitPart::Segment};
        }
    }

    // Joints take absolute priority -- a joint hit always beats a segment hit.
    return bestJointHit.has_value() ? bestJointHit : bestSegHit;
}

void BoneTool::onMouseDown(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& /*imageHitTest*/, float /*hitScale*/,
    bool touchOptimized) {
    scene.beginInteraction();
    const Radii radii = radiiFor(touchOptimized);
    const auto hit =
        hitTestBonePart(input.screenPosition, input.viewSize, scene.skeleton, input.camera, radii.jointRadius,
                        radii.lineRadius);

    if (hit.has_value()) {
        if (input.isCommandPressed || input.isShiftPressed) {
            // Cmd-click toggles the bone in the multi-selection without
            // starting a drag -- matches macOS finder/list conventions.
            scene.toggleBoneSelection(hit->id);
            interaction_ = std::nullopt;
            currentPreview_ = std::nullopt;
            scene.setBoneCreationPreview(std::nullopt, std::nullopt);
            return;
        }
        scene.selectBone(hit->id);
        currentPreview_ = std::nullopt;
        interaction_ = interactionForHit(*hit, input);
        mouseDownPosition_ = input.position;
        moveThresholdExceeded_ = false;
        if (interaction_.has_value() && interaction_->kind == InteractionKind::Create) {
            currentPreview_ = input.position;
            scene.setBoneCreationPreview(interaction_->start, input.position);
        } else {
            scene.setBoneCreationPreview(std::nullopt, std::nullopt);
        }
        return;
    }

    const std::optional<Uuid> chainParentID = scene.selectedBoneID;
    Interaction create;
    create.kind = InteractionKind::Create;
    create.parentID = chainParentID;
    create.start = input.position;
    interaction_ = create;
    currentPreview_ = input.position;
    mouseDownPosition_ = input.position;
    moveThresholdExceeded_ = true; // create mode always tracks drag from frame 0.
    scene.setBoneCreationPreview(input.position, input.position);
}

void BoneTool::onMouseDrag(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& /*imageHitTest*/, float /*hitScale*/,
    bool touchOptimized) {
    if (!interaction_.has_value()) return;
    const Radii radii = radiiFor(touchOptimized);

    switch (interaction_->kind) {
        case InteractionKind::Create:
            currentPreview_ = input.position;
            scene.setBoneCreationPreview(interaction_->start, input.position);
            break;
        case InteractionKind::MoveRoot:
            if (!ensureDragStarted(input.position, radii.clickDragThreshold)) return;
            scene.moveBoneRoot(interaction_->boneID, input.position);
            break;
        case InteractionKind::MoveTip:
            if (!ensureDragStarted(input.position, radii.clickDragThreshold)) return;
            scene.moveBoneTip(interaction_->boneID, input.position);
            break;
    }
}

void BoneTool::onMouseUp(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& /*imageHitTest*/, float /*hitScale*/,
    bool /*touchOptimized*/) {
    scene.endInteraction();

    const std::optional<Interaction> interaction = interaction_;
    const std::optional<Vec2> currentPreview = currentPreview_;

    interaction_ = std::nullopt;
    currentPreview_ = std::nullopt;
    mouseDownPosition_ = std::nullopt;
    moveThresholdExceeded_ = false;
    scene.setBoneCreationPreview(std::nullopt, std::nullopt);

    if (!interaction.has_value() || interaction->kind != InteractionKind::Create) return;
    const Vec2 end = currentPreview.has_value() ? *currentPreview : input.position;

    // A CLICK ON EMPTY CANVAS: a press and release that never travelled far
    // enough to be a bone. The same 8-unit threshold decides both "too
    // short to be a bone" and "a click that cancels" -- breaking the chain
    // (clearing selection) rather than leaving the next drag to chain off
    // whatever was selected before it.
    if (length(end - interaction->start) < 8.0f) {
        scene.selectBone(std::nullopt);
        return;
    }
    scene.addBone(interaction->start, end, interaction->parentID);
}

} // namespace umeshcore
